src/ejson2/EjsonEncoder.ts

```diff
@@ -0,0 +1,593 @@
+    if (this.options.canonical) {
+      if (Number.isInteger(value)) {
+        // Determine if it fits in Int32 or needs Int64
+        if (value >= -2147483648 && value <= 2147483647) {
+          this.writeNumberIntWrapper(value);
+        } else {
+          this.writeNumberLongWrapper(value);
+        }
+      } else {
+        this.writeNumberDoubleWrapper(value);
+      }
+    } else {
+      // Relaxed format
+      if (!isFinite(value)) {
+        this.writeNumberDoubleWrapper(value);
+      } else {
+        this.writeNumber(value);
+      }
+    }
+  }
+
+  private writeNumberIntWrapper(value: number): void {
+    // Write {"$numberInt":"value"}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x496e7422); // "$numberInt"
+    writer.u8(0x3a); // :
+    this.writeStr(value.toString());
+    writer.u8(0x7d); // }
+  }
+
+  private writeNumberLongWrapper(value: number): void {
+    // Write {"$numberLong":"value"}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x4c6f6e67); writer.u8(0x22); // "$numberLong"
+    writer.u8(0x3a); // :
+    this.writeStr(value.toString());
+    writer.u8(0x7d); // }
+  }
+
+  private writeNumberDoubleWrapper(value: number): void {
+    // Write {"$numberDouble":"value"}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x446f7562); writer.u16(0x6c65); writer.u8(0x22); // "$numberDouble"
+    writer.u8(0x3a); // :
+    if (!isFinite(value)) {
+      this.writeStr(this.formatNonFinite(value));
+    } else {
+      this.writeStr(value.toString());
+    }
+    writer.u8(0x7d); // }
+  }
+
+  private writeDateAsEjson(value: Date): void {
+    const timestamp = value.getTime();
+    // Check if date is valid
+    if (isNaN(timestamp)) {
+      throw new Error('Invalid Date');
+    }
+    
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    writer.u32(0x22246461); writer.u16(0x7465); writer.u8(0x22); // "$date"
+    writer.u8(0x3a); // :
+    
+    if (this.options.canonical) {
+      // Write {"$numberLong":"timestamp"}
+      writer.u8(0x7b); // {
+      writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x4c6f6e67); writer.u8(0x22); // "$numberLong"
+      writer.u8(0x3a); // :
+      this.writeStr(timestamp.toString());
+      writer.u8(0x7d); // }
+    } else {
+      // Use ISO format for dates between 1970-9999 in relaxed mode
+      const year = value.getFullYear();
+      if (year >= 1970 && year <= 9999) {
+        this.writeStr(value.toISOString());
+      } else {
+        // Write {"$numberLong":"timestamp"}
+        writer.u8(0x7b); // {
+        writer.u32(0x22246e75); writer.u32(0x6d626572); writer.u32(0x4c6f6e67); writer.u8(0x22); // "$numberLong"
+        writer.u8(0x3a); // :
+        this.writeStr(timestamp.toString());
```

Comment:
Cast to strings by concatenating `""`, throughout the code:

```suggestion
        this.writeStr(timestamp + '');
```